File: results/2026-04-02_163146/tasks/02-log-file-analyzer/default-opus/generated-code/log_analyzer.js
#!/usr/bin/env node
/**
 * Log file analyzer — parses mixed syslog and JSON log formats, extracts
 * errors and warnings, and produces frequency tables with occurrence times.
 *
 * Developed using red/green TDD (see test_log_analyzer.js for the tests
 * that drove each function into existence).
 *
 * Usage:
 *   node log_analyzer.js <logfile> [output.json]
 */

const fs = require("node:fs");
const path = require("node:path");

// ---------------------------------------------------------------------------
// Cycle 1 (GREEN): Parse syslog-style lines
// Pattern: "2024-01-15 08:23:45 ERROR [auth] Some message text"
// ---------------------------------------------------------------------------
const SYSLOG_RE =
  /^(\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2})\s+(ERROR|WARNING|INFO|DEBUG|CRITICAL)\s+\[(\w+)\]\s+(.+)$/;

/**
 * Parse a syslog-format log line into a structured object.
 * Returns null if the line doesn't match the expected pattern.
 */
function parseSyslogLine(line) {
  const trimmed = line.trim();
  const m = trimmed.match(SYSLOG_RE);
  if (!m) return null;
  return {
    timestamp: m[1],
    level: m[2],
    source: m[3],
    message: m[4],
  };
}

// ---------------------------------------------------------------------------
// Cycle 2 (GREEN): Parse JSON-structured lines
// ---------------------------------------------------------------------------

/**
 * Normalise various timestamp formats to "YYYY-MM-DD HH:MM:SS".
 */
function normaliseTimestamp(ts) {
  // Handle ISO-8601 with Z or timezone offset
  // e.g. "2024-01-15T08:35:22Z" → "2024-01-15 08:35:22"
  const isoMatch = ts.match(
    /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})/
  );
  if (isoMatch) {
    return `${isoMatch[1]}-${isoMatch[2]}-${isoMatch[3]} ${isoMatch[4]}:${isoMatch[5]}:${isoMatch[6]}`;
  }
  return ts;
}

/**
 * Parse a JSON-structured log line into a normalised object.
 * Returns null if the line is not valid JSON or lacks required fields.
 */
function parseJsonLine(line) {
  const trimmed = line.trim();
  if (!trimmed.startsWith("{")) return null;

  let data;
  try {
    data = JSON.parse(trimmed);
  } catch {
    return null;
  }

  // Require the four mandatory fields
  for (const key of ["timestamp", "level", "source", "message"]) {
    if (!(key in data)) return null;
  }

  return {
    timestamp: normaliseTimestamp(data.timestamp),
    level: data.level.toUpperCase(),
    source: data.source,
    message: data.message,
  };
}

// ---------------------------------------------------------------------------
// Cycle 3 (GREEN): Unified line parser — tries JSON first, then syslog
// ---------------------------------------------------------------------------

/**
 * Try both parsers; return the first successful result or null.
 */
function parseLine(line) {
  return parseJsonLine(line) || parseSyslogLine(line);
}

// ---------------------------------------------------------------------------
// Cycle 4 (GREEN): Filter only ERROR and WARNING entries
// ---------------------------------------------------------------------------

/**
 * Keep only entries whose level is ERROR or WARNING.
 */
function filterErrorsAndWarnings(entries) {
  return entries.filter(
    (e) => e.level === "ERROR" || e.level === "WARNING"
  );
}

// ---------------------------------------------------------------------------
// Cycle 5 (GREEN): Build frequency table with first/last timestamps
// ---------------------------------------------------------------------------

/**
 * Aggregate entries by (source, message) and track count + timestamps.
 * Returns an array sorted by count descending, then first_seen ascending.
 */
function buildFrequencyTable(entries) {
  const stats = new Map();

  for (const entry of entries) {
    const key = `${entry.source}\0${entry.message}`;

    if (!stats.has(key)) {
      stats.set(key, {
        source: entry.source,
        message: entry.message,
        level: entry.level,
        count: 0,
        first_seen: entry.timestamp,
        last_seen: entry.timestamp,
      });
    }

    const record = stats.get(key);
    record.count += 1;
    if (entry.timestamp < record.first_seen)
      record.first_seen = entry.timestamp;
    if (entry.timestamp > record.last_seen)
      record.last_seen = entry.timestamp;
  }

  return Array.from(stats.values()).sort((a, b) => {
    if (b.count !== a.count) return b.count - a.count;
    return a.first_seen < b.first_seen ? -1 : 1;
  });
}

// ---------------------------------------------------------------------------
// Cycle 6 (GREEN): Parse a full log file from disk
// ---------------------------------------------------------------------------

/**
 * Read a log file line-by-line, parse each line, skip unparseable ones.
 * Throws a clear error if the file doesn't exist.
 */
function parseLogFile(filePath) {
  if (!fs.existsSync(filePath)) {
    throw new Error(`Log file not found: ${filePath}`);
  }

  const content = fs.readFileSync(filePath, "utf-8");
  const lines = content.split("\n");
  const entries = [];

  for (const line of lines) {
    if (line.trim() === "") continue;
    const entry = parseLine(line);
    if (entry !== null) {
      entries.push(entry);
    }
  }

  return entries;
}

// ---------------------------------------------------------------------------
// Cycle 7 (GREEN): Render a human-readable table
// ---------------------------------------------------------------------------

/**
 * Format the frequency table as an aligned, bordered ASCII table.
 */
function formatTable(freqTable) {
  if (freqTable.length === 0) {
    return "No error or warning entries found.";
  }

  const headers = ["Level", "Source", "Message", "Count", "First Seen", "Last Seen"];

  // Build rows as string arrays
  const rows = freqTable.map((row) => [
    row.level,
    row.source,
    row.message,
    String(row.count),
    row.first_seen,
    row.last_seen,
  ]);

  // Compute column widths
  const widths = headers.map((h, i) =>
    Math.max(h.length, ...rows.map((r) => r[i].length))
  );

  const sep =
    "+-" + widths.map((w) => "-".repeat(w)).join("-+-") + "-+";
  const fmtRow = (cells) =>
    "| " +
    cells.map((c, i) => c.padEnd(widths[i])).join(" | ") +
    " |";

  const lines = [sep, fmtRow(headers), sep];
  for (const row of rows) {
    lines.push(fmtRow(row));
  }
  lines.push(sep);

  return lines.join("\n");
}

// ---------------------------------------------------------------------------
// Cycle 8 (GREEN): Full analysis pipeline
// ---------------------------------------------------------------------------

/**
 * Run the complete analysis: parse → filter → frequency table → output.
 * Optionally writes JSON results to jsonOutputPath.
 */
function analyzeLogFile(filePath, jsonOutputPath) {
  const entries = parseLogFile(filePath);
  const filtered = filterErrorsAndWarnings(entries);
  const freq = buildFrequencyTable(filtered);

  const result = {
    file: filePath,
    total_lines_parsed: entries.length,
    errors_and_warnings: filtered.length,
    unique_error_types: freq.length,
    frequency_table: freq,
  };

  if (jsonOutputPath) {
    fs.writeFileSync(jsonOutputPath, JSON.stringify(result, null, 2), "utf-8");
  }

  return result;
}

// ---------------------------------------------------------------------------
// CLI entry point
// ---------------------------------------------------------------------------

function main() {
  const args = process.argv.slice(2);

  if (args.length < 1) {
    console.error("Usage: node log_analyzer.js <logfile> [output.json]");
    process.exit(1);
  }

  const logPath = args[0];
  const jsonPath = args[1] || null;

  let result;
  try {
    result = analyzeLogFile(logPath, jsonPath);
  } catch (err) {
    console.error(`Error: ${err.message}`);
    process.exit(1);
  }

  // Print human-readable output
  console.log(`\nLog Analysis: ${logPath}`);
  console.log(`Total entries parsed: ${result.total_lines_parsed}`);
  console.log(`Errors/Warnings found: ${result.errors_and_warnings}`);
  console.log(`Unique error types: ${result.unique_error_types}`);
  console.log();
  console.log(formatTable(result.frequency_table));

  if (jsonPath) {
    console.log(`\nJSON output written to: ${jsonPath}`);
  }
}

// Export for testing
module.exports = {
  parseSyslogLine,
  parseJsonLine,
  parseLine,
  filterErrorsAndWarnings,
  buildFrequencyTable,
  parseLogFile,
  formatTable,
  analyzeLogFile,
};

// Run CLI if invoked directly
if (require.main === module) {
  main();
}

File: results/2026-04-07_225702/tasks/02-log-file-analyzer/default-opus/generated-code/log_analyzer.py
"""
Log File Analyzer
─────────────────
Parses mixed-format log files (syslog + JSON), extracts errors/warnings,
and produces a frequency table with first/last occurrence timestamps.

Outputs both a human-readable table (stdout) and a JSON report file.
"""

import re
import json
import sys
from collections import defaultdict
from pathlib import Path


# Syslog pattern: "Mon DD HH:MM:SS hostname process[pid]: LEVEL: message"
SYSLOG_RE = re.compile(
    r'^(?P<timestamp>\w{3}\s+\d{1,2}\s+\d{2}:\d{2}:\d{2})\s+'  # timestamp
    r'\S+\s+'                                                      # hostname
    r'\S+:\s+'                                                     # process[pid]:
    r'(?P<level>ERROR|WARNING|INFO|DEBUG|CRITICAL):\s+'            # level
    r'(?P<message>.+)$'                                            # message
)


def parse_syslog_line(line):
    """Parse a syslog-style log line. Returns dict with level/message/timestamp, or None."""
    m = SYSLOG_RE.match(line.strip())
    if not m:
        return None
    return {
        "timestamp": m.group("timestamp"),
        "level": m.group("level"),
        "message": m.group("message"),
    }


def parse_json_line(line):
    """Parse a JSON-structured log line. Returns dict with level/message/timestamp, or None."""
    try:
        data = json.loads(line.strip())
    except (json.JSONDecodeError, ValueError):
        return None
    # Require the three essential fields
    if not all(k in data for k in ("level", "message", "timestamp")):
        return None
    return {
        "timestamp": data["timestamp"],
        "level": data["level"],
        "message": data["message"],
    }


def parse_line(line):
    """Try syslog first, then JSON. Returns parsed dict or None for unparseable lines."""
    return parse_syslog_line(line) or parse_json_line(line)


# Only these levels are extracted for analysis
_ALERT_LEVELS = {"ERROR", "WARNING", "CRITICAL"}


def parse_log_file(path):
    """Read a log file and return a list of parsed error/warning entries.

    Raises FileNotFoundError for missing files.
    """
    path = Path(path)
    if not path.exists():
        raise FileNotFoundError(f"Log file not found: {path}")

    entries = []
    with open(path, encoding="utf-8", errors="replace") as f:
        for line in f:
            parsed = parse_line(line)
            if parsed and parsed["level"] in _ALERT_LEVELS:
                entries.append(parsed)
    return entries


def build_frequency_table(entries):
    """Build a frequency table keyed by error message.

    Each entry in the returned dict has: count, level, first_seen, last_seen.
    Entries are processed in order, so first_seen/last_seen reflect log order.
    """
    table = {}
    for entry in entries:
        msg = entry["message"]
        if msg not in table:
            table[msg] = {
                "level": entry["level"],
                "count": 0,
                "first_seen": entry["timestamp"],
                "last_seen": entry["timestamp"],
            }
        table[msg]["count"] += 1
        table[msg]["last_seen"] = entry["timestamp"]
    return table


def format_table(table):
    """Render the frequency table as a human-readable text table, sorted by count descending."""
    if not table:
        return "No errors or warnings found."

    # Sort by count descending
    sorted_items = sorted(table.items(), key=lambda kv: kv[1]["count"], reverse=True)

    # Calculate column widths dynamically
    headers = ("Message", "Level", "Count", "First Seen", "Last Seen")
    rows = [
        (msg, info["level"], str(info["count"]), info["first_seen"], info["last_seen"])
        for msg, info in sorted_items
    ]
    widths = [
        max(len(h), max(len(r[i]) for r in rows))
        for i, h in enumerate(headers)
    ]

    sep = "+-" + "-+-".join("-" * w for w in widths) + "-+"
    fmt = "| " + " | ".join(f"{{:<{w}}}" for w in widths) + " |"

    lines = [sep, fmt.format(*headers), sep]
    for row in rows:
        lines.append(fmt.format(*row))
    lines.append(sep)
    return "\n".join(lines)


def write_json_report(table, output_path):
    """Write the frequency table as a structured JSON report file."""
    sorted_items = sorted(table.items(), key=lambda kv: kv[1]["count"], reverse=True)

    report = {
        "summary": {
            "total_entries": sum(info["count"] for info in table.values()),
            "unique_messages": len(table),
        },
        "errors": [
            {
                "message": msg,
                "level": info["level"],
                "count": info["count"],
                "first_seen": info["first_seen"],
                "last_seen": info["last_seen"],
            }
            for msg, info in sorted_items
        ],
    }

    output_path = Path(output_path)
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(report, f, indent=2)
    return output_path


def analyze(log_path, json_output_path):
    """Core analysis pipeline: parse -> frequency table -> output both formats.

    Prints the human-readable table to stdout and writes the JSON report.
    Calls sys.exit(1) with an error message if the log file is not found.
    """
    try:
        entries = parse_log_file(log_path)
    except FileNotFoundError as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)

    table = build_frequency_table(entries)
    print(format_table(table))

    if table:
        write_json_report(table, json_output_path)
        print(f"\nJSON report written to: {json_output_path}")
    else:
        # Write an empty report for consistency
        write_json_report(table, json_output_path)


def main():
    """CLI entry point. Usage: python log_analyzer.py <logfile> [output.json]"""
    if len(sys.argv) < 2:
        print("Usage: python log_analyzer.py <logfile> [output.json]", file=sys.stderr)
        sys.exit(1)

    log_path = Path(sys.argv[1])
    json_output = Path(sys.argv[2]) if len(sys.argv) > 2 else log_path.with_suffix(".report.json")
    analyze(log_path, json_output)


if __name__ == "__main__":
    main()
